pkg/download/download.go

```diff
@@ -75,7 +77,39 @@ func (p *protocol) request(mod, ver string) error {
 }
 
 func (p *protocol) List(ctx context.Context, mod string) ([]string, error) {
-	return p.dp.List(ctx, mod)
+	const op errors.Op = "protocol.List"
+	goList, goErr := p.dp.List(ctx, mod)
+	strList, sErr := p.s.List(ctx, mod)
+	if goErr != nil && sErr != nil {
```

Comment:
We should return if either of errors are nil, not both. Otherwise, the results will not be consistent to the user.

Reply:
@marwan-at-work could you expand on that?
There are 4 cases:
1. `go list` failed (i.e. repo has been deleted like in the issue #590 ) and storage has the versions -> versions from storage
2. `go list` succeeded and storage failed (i.e. is empty) -> versions from `go list`
3. both `go list` and storage returned their lists -> combined lists storage + `go list`
4. `go list` and storage failed -> return error

Or did you mean we should differentiate between ErrNotFound and unexpected error? We could do that for the storage but I don't know how to do that for `go list`...

Reply:
@marpio 
1. Can we gracefully handle that error? If a repo is deleted, `go list -m` should return a 404 figuratively speaking. But if let's say github is down, then that's an unexpected error. Do we want to ignore upstream if upstream is unavailable? Obviously we ignore upstream if the "client" already knows what version it wants to get from Athens. Because `/list` never gets called. 
2. If storage fails, we should fail. always. 
3. this is happy path. We're good. 
4. Since storage failed, this one fails as well. Don't think we need to combine the errors, just take the storage error: that's more important. But up to you.

Reply:
My answer is that, we should be able to ask `go list -m` questions as such: 
- Athens: why did you error out? 
- Go: repository not found 
- Athens: ok, this is not an error, only take from storage. 

- Athens: why did you error out? 
- Go: GitHub is down
- Athens: ok this is an error, we should report it to our metrics and tell the user to try again. Because we might give them an outdated list. Of course, /list is now down but /info /mod and /zip still work fine. 

And so on and so forth. The biggest win from this is that Athens remains reproducible. If we just simply ignore upstream errors, then one second /list returns a full list, the next second returns only the storage lists. And we'll never be able to make Athens stable in that regard. 

Parsing the `go list -m` errors is gonna be ugly, see this comment: 
https://github.com/gomods/athens/issues/590#issuecomment-418020244

I hope the go command can in the future return structured errors.

Reply:
Definitely makes sense. Thank you Marwan!